sorc/chgres_cube.fd/program_setup.F90

```diff
@@ -88,13 +88,15 @@ module program_setup
  integer, public                 :: cycle_hour = -999 !< Cycle hour.
  integer, public                 :: regional = 0 !<  For regional target grids.  When '1' remove boundary halo region from atmospheric/surface data and
                                                  !! output atmospheric boundary file. When '2' output boundary file only. Default is '0' (global grids).
+ integer, public                 :: calrh = 0 !< rh to spfh calculation. 0:original; 1: GFSv15/16                                         
```

Comment:
Instead of setting two new flags - calrh and use_rh - can this option be communicated thru the varmap table somehow?

Reply:
I don't know how to do that. These two parameters are only for humidity and not for other tracers/variables. Not sure if we could add extra columns only for one particular variable in varmap table. I wonder if @LarissaReames-NOAA and @JeffBeck-NOAA have any thoughts on this.

Reply:
After discussing with @LarissaReames-NOAA, we believe that "calrh" and "use_rh" should stay as namelist entries, since the varmap table is really for handling of fields that are required by the CCPP scheme the user wishes to run, while this is a variable calculation option, which depends on how someone wishes to initialize.

Reply:
What if we make these namelist options for super users who wish to have complete control, but default to what we think is the "best practice" options based on data type, external model, and data date? So in this case, switch to the new calculation type for grib2, GFS, post-whatever date the new GFS data was implemented in the grib2 files. This would keep the vast majority of users from ever having to worry about specifying these options.

Reply:
@LarissaReames-NOAA good idea. But I am not sure which option fits most scenarios. I have some recommendations as follows:

For grib2 full level qv, use qv (e.g. GFSv15 combined, GFSv16 uncombined) 
For grib2 qv with missing levels or not available, use rh with consistent conversion (e.g. GFSv15 uncombined, GFSv16 combined)

As for specific GFS grib2 data: 
GFSv15 uncombined: use_rh=True; calrh=1 
GFSv15 combined: use_rh=False 
GFSv16 uncombined: use_rh=False 
GFSv16 combined: use_rh=True; calrh=1

Maybe we can set default as "use_rh=True; calrh=1" since this may be the best option for the current operational GFS grib2 (combined).

Reply:
Made a couple of changes:
1. remove calrh and use external_model to determine rh to spfh conversion: if external_model is GFS, rh2spfh_gfs will be used;
2. default use_rh as True as the option for current combined operational GFSV16 grib2 files

Reply:
The changes will need the following regression test baseline update since now the default set up uses RH instead of SPFH for grib2 files:

consistency.log08:<<< C192 GFS GRIB2 TEST FAILED. >>>
consistency.log09:<<< 25-KM CONUS GFS GRIB2 TEST FAILED. >>>
consistency.log10:<<< 3-km CONUS HRRR W/ GFS PHYSICS GRIB2 TEST FAILED. <<<
consistency.log11:<<< 3-km CONUS HRRR W/ GSD PHYSICS AND SFC FROM FILE GRIB2 TEST FAILED. <<<
consistency.log13:<<< 13-km CONUS RAP W/ GSD PHYSICS AND SFC FROM FILE GRIB2 TEST FAILED. <<<
consistency.log14:<<< 13-KM NA GFS NCEI GRIB2 TEST FAILED. >>>
consistency.log16:<<< 25-KM CONUS GFS PGRIB2+BGRIB2 TEST FAILED. >>>

Please let me know how you would like me to proceed.

Reply:
> Made a couple of changes:
> 
> 1. remove calrh and use external_model to determine rh to spfh conversion: if external_model is GFS, rh2spfh_gfs will be used;
Is this really the correct calculate for *all* GFS data, or just for dates after the switch to V15?
> 2. default use_rh as True as the option for current combined operational GFSV16 grib2 files
So the majority of files that chgres_cube is capable of reading do actually have spfh, so I think it makes much more sense to have use of spfh be the default.

As for your failed tests, I certainly feel like if the files have all levels of spfh already, and not rh, but you've chosen to use RH, then the code should quit and provide an error. The functionalities that you added should not be breaking RAP and HRRR consistency tests since your code is not relevant to these files at all.

Reply:
@JiliDong-NOAA, since all of this is specific to the GFS, I think we should only be modifying "RH" vs "SPFH" for GFS grib2 data based on date, therefore RAP- and HRRR-based grib2 file processing should stay the same.

Also, there seems to be a discrepancy here:

As for specific GFS grib2 data:
GFSv15 uncombined: use_rh=True; calrh=1
GFSv15 combined: use_rh=False
GFSv16 uncombined: use_rh=False
GFSv16 combined: use_rh=True; calrh=1

Use of "RH" is flipped for uncombined vs. combined from v15 to v16?

Reply:
I made another change so that using rh or spfh in grib2 will be based on if spfh are on the same level as UGRD: if not, rh will be used. "use_rh" is also removed. The current code doesn't have any additional parameters. The regression test baseline needs to be updated for GFS grib2:

consistency.log08:<<< C192 GFS GRIB2 TEST FAILED. >>>
consistency.log09:<<< 25-KM CONUS GFS GRIB2 TEST FAILED. >>>
consistency.log14:<<< 13-KM NA GFS NCEI GRIB2 TEST FAILED. >>>
consistency.log16:<<< 25-KM CONUS GFS PGRIB2+BGRIB2 TEST FAILED. >>>

Reply:
> > Made a couple of changes:
> > 
> > 1. remove calrh and use external_model to determine rh to spfh conversion: if external_model is GFS, rh2spfh_gfs will be used;
> > 2. default use_rh as True as the option for current combined operational GFSV16 grib2 files
> 
> 1: Is this really the correct calculation for _all_ GFS data, or just for dates after the switch to V15?
> 
> 2: So the majority of files that chgres_cube is capable of reading do actually have spfh, so I think it makes much more sense to have use of spfh be the default.
> 
> As for your failed tests, I certainly feel like if the files have all levels of spfh already, and not rh, but you've chosen to use RH, then the code should quit and provide an error. The functionalities that you added should not be breaking RAP and HRRR consistency tests since your code is not relevant to these files at all.

rh2spfh_gfs is for GFSv15/16 or FV3GFS.

Reply:
> @JiliDong-NOAA, since all of this is specific to the GFS, I think we should only be modifying "RH" vs "SPFH" for GFS grib2 data based on date, therefore RAP- and HRRR-based grib2 file processing should stay the same.
> 
> Also, there seems to be a discrepancy here:
> 
> As for specific GFS grib2 data:
> GFSv15 uncombined: use_rh=True; calrh=1
> GFSv15 combined: use_rh=False
> GFSv16 uncombined: use_rh=False
> GFSv16 combined: use_rh=True; calrh=1
> 
> Use of "RH" is flipped for uncombined vs. combined from v15 to v16?

After the most recent change, HRRR/RAP processing should stay the same if SPFH are on the same levels as UGRD. For the last question, it is not flipped. GFS v15 uncombined grib2 has no spfh so rh is used. GFSv16 combined has missing levels of spfh and rh is preferred. Both GFSv15 combined and GFSv16 uncombined has spfh on the same levels as UGRD and are considered as full levels spfh.

Reply:
I just confirmed with Fanglin that all GFS versions use the same conversion of spfh to rh.

Reply:
The current code update allows chgres to:

- automatically choose reading SPFH or RH based on whether grib2 data contains SPFH at full levels as UGRD
- automatically choose which rh to spfh conversions based on external_model (GFS vs. non-GFS)

The previous added parameters (use_rh and calrh) are removed.